lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -377,6 +456,29 @@ def _module_file_from_path(leaf_name, path):
 
         return module_path, has_code, package_path
 
+    def exec_module(self, module):
+        # execute the module's code in its namespace
+        code_obj = self.get_code(self._fullname)
+        if code_obj is not None:  # things like NS packages that can't have code on disk will return None
+            exec(code_obj, module.__dict__)
+
+    def create_module(self, spec):
+        # short-circuit redirect; we've already imported the redirected module, so just alias it and return it
+        if self._redirect_module:
+            sys.modules[self._fullname] = self._redirect_module
+            return self._redirect_module
+
+        module = ModuleType(spec.name)
+        module.__loader__ = spec.loader
```

Comment:
Reading pep489 I also see this in regards to the creation phase:

```
This function is not responsible for setting import-related attributes specified in PEP 451 [1] (such as __name__ or __loader__) on the new module.
```

pep451 (https://www.python.org/dev/peps/pep-0451/#attributes) says:

> The reverse of this mapping describes how the import machinery sets the module attributes right before calling exec_module().

So I'm thinking we shouldn't be doing most of this manipulation of `module` other than just creating the module object.

Sorry, I'm beginning to ramble as I think this through...

But based on this and other comments, I question whether we need `create_module` at all.  Or rather, we just have it return `None`.

Reply:
Yep

Reply:
Fixed